File: eShopLegacyMVCSolution/src/eShopLegacyMVC-SpringBoot/src/main/java/com/eshop/catalog/controller/CatalogController.java
package com.eshop.catalog.controller;

import com.eshop.catalog.model.CatalogItem;
import com.eshop.catalog.service.ICatalogService;

import jakarta.validation.Valid;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.validation.BindingResult;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.ModelAttribute;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.server.ResponseStatusException;

@Controller
@RequestMapping("/catalog")
public class CatalogController {

    private static final Logger log = LoggerFactory.getLogger(CatalogController.class);

    private final ICatalogService service;

    public CatalogController(ICatalogService service) {
        this.service = service;
    }

    @GetMapping
    public String index(@RequestParam(defaultValue = "10") int pageSize,
                        @RequestParam(defaultValue = "0") int pageIndex,
                        Model model) {
        log.info("Now loading... /Catalog/Index?pageSize={}&pageIndex={}", pageSize, pageIndex);
        var paginatedItems = service.getCatalogItemsPaginated(pageSize, pageIndex);
        paginatedItems.getData().forEach(item ->
                item.setPictureUri("/items/" + item.getId() + "/pic"));
        model.addAttribute("model", paginatedItems);
        return "catalog/index";
    }

    @GetMapping("/details/{id}")
    public String details(@PathVariable Integer id, Model model) {
        log.info("Now loading... /Catalog/Details?id={}", id);
        if (id == null) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST);
        }
        CatalogItem catalogItem = service.findCatalogItem(id);
        if (catalogItem == null) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND);
        }
        catalogItem.setPictureUri("/items/" + catalogItem.getId() + "/pic");
        model.addAttribute("catalogItem", catalogItem);
        return "catalog/details";
    }

    @GetMapping("/create")
    public String createForm(Model model) {
        log.info("Now loading... /Catalog/Create");
        model.addAttribute("brands", service.getCatalogBrands());
        model.addAttribute("types", service.getCatalogTypes());
        model.addAttribute("catalogItem", new CatalogItem());
        return "catalog/create";
    }

    @PostMapping("/create")
    public String create(@Valid @ModelAttribute CatalogItem catalogItem,
                         BindingResult result,
                         Model model) {
        log.info("Now processing... /Catalog/Create?catalogItemName={}", catalogItem.getName());
        if (result.hasErrors()) {
            model.addAttribute("brands", service.getCatalogBrands());
            model.addAttribute("types", service.getCatalogTypes());
            return "catalog/create";
        }
        service.createCatalogItem(catalogItem);
        return "redirect:/catalog";
    }

    @GetMapping("/edit/{id}")
    public String editForm(@PathVariable Integer id, Model model) {
        log.info("Now loading... /Catalog/Edit?id={}", id);
        if (id == null) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST);
        }
        CatalogItem catalogItem = service.findCatalogItem(id);
        if (catalogItem == null) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND);
        }
        catalogItem.setPictureUri("/items/" + catalogItem.getId() + "/pic");
        model.addAttribute("brands", service.getCatalogBrands());
        model.addAttribute("types", service.getCatalogTypes());
        model.addAttribute("catalogItem", catalogItem);
        return "catalog/edit";
    }

    @PostMapping("/edit/{id}")
    public String edit(@Valid @ModelAttribute CatalogItem catalogItem,
                       BindingResult result,
                       Model model) {
        log.info("Now processing... /Catalog/Edit?id={}", catalogItem.getId());
        if (result.hasErrors()) {
            model.addAttribute("brands", service.getCatalogBrands());
            model.addAttribute("types", service.getCatalogTypes());
            return "catalog/edit";
        }
        service.updateCatalogItem(catalogItem);
        return "redirect:/catalog";
    }

    @GetMapping("/delete/{id}")
    public String deleteConfirm(@PathVariable Integer id, Model model) {
        log.info("Now loading... /Catalog/Delete?id={}", id);
        if (id == null) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST);
        }
        CatalogItem catalogItem = service.findCatalogItem(id);
        if (catalogItem == null) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND);
        }
        catalogItem.setPictureUri("/items/" + catalogItem.getId() + "/pic");
        model.addAttribute("catalogItem", catalogItem);
        return "catalog/delete";
    }

    @PostMapping("/delete/{id}")
    public String delete(@PathVariable Integer id) {
        log.info("Now processing... /Catalog/DeleteConfirmed?id={}", id);
        CatalogItem catalogItem = service.findCatalogItem(id);
        service.removeCatalogItem(catalogItem);
        return "redirect:/catalog";
    }
}

File: src/eShop.Web/Controllers/CatalogController.cs
using eShop.Catalog.Data;
using eShop.Catalog.Domain;
using eShop.Web.Configuration;
using eShop.Web.ViewModels;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Options;

namespace eShop.Web.Controllers;

/// <summary>
/// Port of the legacy MVC 5 <c>eShopLegacyMVC.Controllers.CatalogController</c>: same routes, same
/// anti-forgery placement, same 400/404 argument handling and the same redirects to the catalog
/// index (which is the application root).
/// </summary>
public class CatalogController : Controller
{
    /// <summary>Properties the Create and Edit posts are allowed to bind (legacy [Bind(Include = …)]).</summary>
    private const string EditableProperties =
        "Id,Name,Description,Price,PictureFileName,CatalogTypeId,CatalogBrandId,AvailableStock,RestockThreshold,MaxStockThreshold,OnReorder";

    private readonly ICatalogService _service;
    private readonly ILogger<CatalogController> _log;
    private readonly CatalogWebOptions _options;

    public CatalogController(
        ICatalogService service,
        IOptions<CatalogWebOptions> options,
        ILogger<CatalogController> log)
    {
        _service = service;
        _options = options.Value;
        _log = log;
    }

    // GET /[?pageSize=3&pageIndex=10]
    public IActionResult Index(int pageSize = 10, int pageIndex = 0)
    {
        _log.LogInformation("Now loading... /Catalog/Index?pageSize={PageSize}&pageIndex={PageIndex}", pageSize, pageIndex);
        var paginatedItems = _service.GetCatalogItemsPaginated(pageSize, pageIndex);
        ChangeUriPlaceholder(paginatedItems.Data);

        return View(paginatedItems);
    }

    // GET: Catalog/Details/5
    public IActionResult Details(int? id)
    {
        _log.LogInformation("Now loading... /Catalog/Details?id={Id}", id);
        if (id is null)
        {
            return BadRequest();
        }

        var catalogItem = _service.FindCatalogItem(id.Value);
        if (catalogItem is null)
        {
            return NotFound();
        }

        AddUriPlaceHolder(catalogItem);

        return View(catalogItem);
    }

    // GET: Catalog/Create
    public IActionResult Create()
    {
        _log.LogInformation("Now loading... /Catalog/Create");

        return View(BuildFormViewModel(new CatalogItem()));
    }

    // POST: Catalog/Create
    [HttpPost]
    [ValidateAntiForgeryToken]
    public IActionResult Create([Bind(EditableProperties)] CatalogItem item)
    {
        _log.LogInformation("Now processing... /Catalog/Create?catalogItemName={Name}", item.Name);
        if (ModelState.IsValid)
        {
            _service.CreateCatalogItem(item);
            return RedirectToAction(nameof(Index));
        }

        return View(BuildFormViewModel(item));
    }

    // GET: Catalog/Edit/5
    public IActionResult Edit(int? id)
    {
        _log.LogInformation("Now loading... /Catalog/Edit?id={Id}", id);
        if (id is null)
        {
            return BadRequest();
        }

        var catalogItem = _service.FindCatalogItem(id.Value);
        if (catalogItem is null)
        {
            return NotFound();
        }

        AddUriPlaceHolder(catalogItem);

        return View(BuildFormViewModel(catalogItem));
    }

    // POST: Catalog/Edit/5
    [HttpPost]
    [ValidateAntiForgeryToken]
    public IActionResult Edit([Bind(EditableProperties)] CatalogItem item)
    {
        _log.LogInformation("Now processing... /Catalog/Edit?id={Id}", item.Id);
        if (ModelState.IsValid)
        {
            _service.UpdateCatalogItem(item);
            return RedirectToAction(nameof(Index));
        }

        AddUriPlaceHolder(item);

        return View(BuildFormViewModel(item));
    }

    // GET: Catalog/Delete/5
    public IActionResult Delete(int? id)
    {
        _log.LogInformation("Now loading... /Catalog/Delete?id={Id}", id);
        if (id is null)
        {
            return BadRequest();
        }

        var catalogItem = _service.FindCatalogItem(id.Value);
        if (catalogItem is null)
        {
            return NotFound();
        }

        AddUriPlaceHolder(catalogItem);

        return View(catalogItem);
    }

    // POST: Catalog/Delete/5
    [HttpPost]
    [ActionName(nameof(Delete))]
    [ValidateAntiForgeryToken]
    public IActionResult DeleteConfirmed(int id)
    {
        _log.LogInformation("Now processing... /Catalog/DeleteConfirmed?id={Id}", id);
        var catalogItem = _service.FindCatalogItem(id);
        if (catalogItem is not null)
        {
            _service.RemoveCatalogItem(catalogItem);
        }

        return RedirectToAction(nameof(Index));
    }

    private CatalogItemFormViewModel BuildFormViewModel(CatalogItem item) =>
        new(item, _service.GetCatalogBrands(), _service.GetCatalogTypes());

    private void ChangeUriPlaceholder(IEnumerable<CatalogItem> items)
    {
        foreach (var catalogItem in items)
        {
            AddUriPlaceHolder(catalogItem);
        }
    }

    /// <summary>
    /// Legacy behaviour: the picture URI is an absolute link to <c>items/{id}/pic</c>. That route
    /// now lives in the catalog API, so the base address is configurable and falls back to the
    /// current request's scheme and host.
    /// </summary>
    private void AddUriPlaceHolder(CatalogItem item)
    {
        var baseUrl = string.IsNullOrWhiteSpace(_options.PicturesBaseUrl)
            ? $"{Request.Scheme}://{Request.Host}"
            : _options.PicturesBaseUrl.TrimEnd('/');

        item.PictureUri = $"{baseUrl}/items/{item.Id}/pic";
    }
}
